core/sdk-core/src/main/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -87,6 +90,64 @@ public URI get(String accessKey, EndpointDiscoveryRequest request) {
         return endpoint.endpoint();
     }
 
+    public CompletableFuture<URI> getAsync(String accessKey, EndpointDiscoveryRequest request) {
+        key = accessKey;
+
+        // Support null (anonymous credentials) by mapping to empty-string. The backing cache does not support null.
+        if (key == null) {
+            key = "";
+        }
+
+        if (request.cacheKey().isPresent()) {
+            key = key + ":" + request.cacheKey().get();
+        }
+
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        if (endpoint == null) {
+            if (request.required()) {
+                return discoverEndpoint(request).handle(
+                    (endpointDiscoveryEndpoint, throwable) -> {
+                        if (throwable != null) {
+                            if (throwable instanceof InterruptedException) {
+                                Thread.currentThread().interrupt();
+                                throw EndpointDiscoveryFailedException.create(throwable);
+                            }
+                            if (throwable instanceof ExecutionException
+                                    || throwable instanceof CompletionException) {
+                                throw EndpointDiscoveryFailedException.create(throwable.getCause());
+                            }
+                            throw new RuntimeException("new exception");
```

Comment:
Added by mistake? Or does this need to wrap the original error/cause?

Reply:
oh this is a mistake. Will change this